### PR TITLE
Multiple paths

### PR DESCRIPTION
**Pull request by SCiarella**
**This PR has the following two main objectives**

(1) Fix a small bug in the definition of the free energy graph:

Instead of weighting the graph edge between `i` and `j` as
`weight = F[j]`
we now do instead 
`weight = 0.5*(F[i] + F[j])`.
This guarantees that the graph is well balanced and the optimal path is calculated corrected in all edge cases.
The tests have been update according to this new definition.

(2) Address #204 

I have added the function `paths.multiple_paths` that returns a list of `Pathway` object, corresponding to the desired number of optimal paths between two selected sites.
The multiple paths are collected according to [this algorithm](https://doi.org/10.1287/mnsc.17.11.712) (implemented in [networkx](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.simple_paths.shortest_simple_paths.html#shortest-simple-paths) ) , while discarding all the ones that are too similar.
I have updated `energy_along_path` and `path_on_landscape` to accept lists of paths as input and produce plots like this one:

### CODE REVIEW

**Comment by SCiarella** [comment]
I have added [this notebook](https://github.com/GEMDAT-repos/gemdat-notebooks/blob/main/multiple_paths.ipynb) to showcase the multiple path functionality.

File: src/gemdat/path.py [hunk 1]
```diff
@@ -187,15 +187,15 @@ def free_energy_graph(F: np.ndarray,
         for move in movements:
             neighbor = tuple((node + move) % F.shape)
             if neighbor in G.nodes:
-                exp_n_energy = np.exp(F[neighbor])
+                exp_n_energy = np.exp(0.5 * (F[node] + F[neighbor]))
```
Comment (stefsmeets) [inline src/gemdat/path.py:190]:
```suggestion
                weight = 0.5 * (F[node] + F[neighbor])
                exp_n_energy = np.exp(weight)
```

File: src/gemdat/path.py [hunk 2]
```diff
@@ -205,6 +205,101 @@ def free_energy_graph(F: np.ndarray,
...
+    Returns
+    -------
+    difference : float
+        Difference between the two paths
+    """
+
+    # Find the shortest and longest paths
+    shortest_path = path_sites1 if len(path_sites1) <= len(
+        path_sites2) else path_sites2
+    longest_path = path_sites2 if shortest_path is path_sites1 else path_sites1
```
Comment (stefsmeets) [inline src/gemdat/path.py:228]: There must be a cleverer way to do this, but just keep it simple for readability 😅

```suggestion
    if len(path1) <= len(path2):
        shortest, longest = path1, path2
    else:
        shortest, longest = path2, path1
```

File: src/gemdat/path.py [hunk 3]
```diff
@@ -205,6 +205,101 @@ def free_energy_graph(F: np.ndarray,
...
+    return 1 - (shared_nodes / len(shortest_path))
+
+
+def multiple_paths(
+    F_graph: nx.Graph,
+    start: tuple,
+    end: tuple,
+    method: _PATHFINDING_METHODS = 'dijkstra',
+    Npaths: int = 3,
+    min_diff: float = 0.15,
```
Comment (stefsmeets) [inline src/gemdat/path.py:245]: Best to force keyword arguments for such functions. This prevents errors when the api changes or if you mix up start/stop at some point.

Another point is to choose good 'opposites' as variable names. Either start<->stop, or begin<->end.

Finally, try to avoid capitals in variable names if you can.

```suggestion
    F_graph: nx.Graph,
    *,
    start: tuple,
    stop: tuple,
    method: _PATHFINDING_METHODS = 'dijkstra',
    n_paths: int = 3,
    min_diff: float = 0.15,
```

Comment (stefsmeets) [reply src/gemdat/path.py:245]: Also can you update the types here for start/stop?

File: src/gemdat/path.py [hunk 4]
```diff
@@ -205,6 +205,101 @@ def free_energy_graph(F: np.ndarray,
...
+        Method used to calculate the shortest path. Options are:
+        - 'dijkstra': Dijkstra's algorithm
+        - 'bellman-ford': Bellman-Ford algorithm
+        - 'minmax-energy': Minmax energy algorithm
+        - 'dijkstra-exp': Dijkstra's algorithm with exponential weights
+    Npaths : int
+        Number of paths to be calculated
+    min_diff : float
+        Minimum difference between the paths
+    """
```
Comment (stefsmeets) [inline src/gemdat/path.py:269]: Return statement is missing in the docstring

File: src/gemdat/path.py [hunk 5]
```diff
@@ -205,6 +205,101 @@ def free_energy_graph(F: np.ndarray,
...
+                    raise ValueError('Path too similar to another')
+        except ValueError:
+            continue
+
+        list_of_paths_sites.append(path)
+        path_energy = [F_graph.nodes[node]['energy'] for node in path]
+        list_of_paths.append(Pathway(sites=path, energy=path_energy))
+
+        if len(list_of_paths) == Npaths:
+            break
```
Comment (stefsmeets) [inline src/gemdat/path.py:298]: I think it is not necessary to keep track of `list_of_paths_sites` here, these are already included in `list_of_paths`.

You can also try to refactor that try/except block into its own function, to make the code a bit nicer to read. Breaking out of 2 loops at once is a bit awkward, and refactoring the inner loop into its own function and using return is a common pattern to do it. 

```suggestion
    list_of_paths = [best_path]

    # Compute the iterator over all the short paths
    all_paths = nx.shortest_simple_paths(F_graph,
                                         source=start,
                                         target=end,
                                         weight='weight')

    # Attempt to find the Np shortest paths
    for idx, path in enumerate(all_paths):
       if path_too_similar(path, list_of_paths):
           continue

        path_energy = [F_graph.nodes[node]['energy'] for node in path]
        list_of_paths.append(Pathway(sites=path, energy=path_energy))

        if len(list_of_paths) == Npaths:
            break
```

File: src/gemdat/plots/matplotlib/_paths.py [hunk 6]
```diff
@@ -7,28 +7,42 @@
 from gemdat.volume import Volume
 
 
-def energy_along_path(*, path: Pathway, structure: Structure,
+def energy_along_path(*, paths: Pathway | list[Pathway], structure: Structure,
```
Comment (stefsmeets) [inline src/gemdat/plots/matplotlib/_paths.py:10]: Could be a clever use for *paths here. Not sure, just a thought.

```suggestion
def energy_along_path(*paths: Pathway | list[Pathway], *, structure: Structure,
```

File: src/gemdat/plots/matplotlib/_paths.py [hunk 7]
```diff
@@ -77,6 +91,16 @@ def energy_along_path(*, path: Pathway, structure: Structure,
                           rotation=45)
     ax_up.get_yaxis().set_visible(False)
 
+    # If available, plot the other pathways
+    if isinstance(paths, list):
+        for idx, path in enumerate(paths[1:]):
+            if path.energy is None:
+                raise ValueError('Pathway does not contain energy data')
+            ax.plot(range(len(path.energy)),
+                    path.energy,
```
Comment (stefsmeets) [inline src/gemdat/plots/matplotlib/_paths.py:100]: Isn't it implied? If you leave out the range, does matplotlib not automatically generate the range?

```suggestion
            ax.plot(path.energy,
```

**Review by stefsmeets — APPROVED** [review]
Cool, nice work! Some python suggestions/nitpicks from my side.

File: src/gemdat/path.py (hunk 3 above)
Comment (SCiarella) [reply src/gemdat/path.py]: done!

File: src/gemdat/path.py (hunk 4 above)
Comment (SCiarella) [reply src/gemdat/path.py:301]: thanks! i missed that!

File: src/gemdat/plots/matplotlib/_paths.py (hunk 6 above)
Comment (SCiarella) [reply src/gemdat/plots/matplotlib/_paths.py:10]: Not sure about this, I think if I prefer *, paths for clarity. Is there a technical advantage that you see?

File: src/gemdat/plots/matplotlib/_paths.py (hunk 7 above)
Comment (SCiarella) [reply src/gemdat/plots/matplotlib/_paths.py:100]: true!

File: src/gemdat/path.py [hunk 8]
```diff
@@ -205,10 +206,130 @@ def free_energy_graph(F: np.ndarray,
...
+    and its implementation in NetworkX. Only paths that are different by at least min_diff are considered.
+
+    Parameters
+    ----------
+    F_graph : nx.Graph
+        Graph of the free energy
+    start : tuple
+        Coordinates of the starting point
+    stop: tuple
+        Coordinates of the stoping point
```
Comment (stefsmeets) [inline src/gemdat/path.py:286]:
```suggestion
        Coordinates of the stopping point
```

File: src/gemdat/plots/matplotlib/_paths.py (hunk 6 above)
Comment (stefsmeets) [reply src/gemdat/plots/matplotlib/_paths.py:10]: The technical advantage is that with `*paths`, the paths variable is always a list, so you don't need to do any instance checking. It's not super common though so it's a bit awkward to use sometimes. I also think `*, paths` is clearer if documented well.

File: src/gemdat/path.py [hunk 9]
```diff
@@ -205,10 +206,130 @@ def free_energy_graph(F: np.ndarray,
...
+    -------
+    difference : float
+        Difference between the two paths
+    """
+
+    # Find the shortest and longest paths
+    if len(path1) <= len(path2):
+        shortest, longest = path1, path2
+    else:
+        shortest, longest = path2, path1
```
Comment (stefsmeets) [inline src/gemdat/path.py:230]: Cool idea with the min/max! I think this should work:

```suggestion
    shortest, longest = sorted((path1,path2), key=len)
```

**Review by stefsmeets — APPROVED** [review]
Looks good to me! 🚀

File: src/gemdat/path.py (hunk 9 above)
Comment (SCiarella) [reply src/gemdat/path.py:230]: This looks neat!
